File: contrail-web-client/src/main/js/Network/component/ClientComponent.js
/*
 * Copyright (C)2012 D. Plaindoux.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation; either version 2, or (at your option) any
 * later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 */

/*global define*/

define([ "Core/object/jObj", "Contrail/jContrail", "./flow/client/ClientComponentDownStreamDataFlow", "./flow/client/ClientComponentUpStreamDataFlow" ],
    function (jObj, jContrail, clientDownStream, clientUpStream) {
        "use strict";

        function ClientComponent(endPoint) {
            jObj.bless(this, jContrail.component.pipeline());

            this.endPoint = endPoint;
            this.upStreamDataFlow = clientUpStream(this);
            this.downStreamDataFlow = clientDownStream(this);

            this.destinations = [];
            this.sources = [];
        }

        ClientComponent.init = jObj.constructor([ jObj.types.Nullable(jObj.types.String) ],
            function (identifier) {
                return new ClientComponent(identifier);
            });

        ClientComponent.prototype.getEndPoint = jObj.method([], jObj.types.Nullable(jObj.types.String),
            function () {
                return this.endPoint;
            });

        ClientComponent.prototype.addDestinationId = jObj.procedure([ jObj.types.String ],
            function (identifier) {
                if (!this.acceptDestinationId(identifier)) {
                    this.destinations.push(identifier);
                }
            });

        ClientComponent.prototype.acceptDestinationId = jObj.method([ jObj.types.String ], jObj.types.Boolean,
            function (identifier) {
                return this.destinations.some(function (id) {
                    return identifier === id;
                });
            });

        ClientComponent.prototype.addSourceId = jObj.procedure([ jObj.types.String ],
            function (identifier) {
                if (!this.acceptSourceId(identifier)) {
                    this.sources.push(identifier);
                }
            });

        ClientComponent.prototype.acceptSourceId = jObj.method([ jObj.types.String ], jObj.types.Boolean,
            function (identifier) {
                return this.sources.some(function (id) {
                    return identifier === id;
                });
            });

        ClientComponent.prototype.getUpStreamDataFlow = jObj.method([], jObj.types.Named("DataFlow"),
            function () {
                return this.upStreamDataFlow;
            });

        ClientComponent.prototype.getDownStreamDataFlow = jObj.method([], jObj.types.Named("DataFlow"),
            function () {
                return this.downStreamDataFlow;
            });

        return ClientComponent.init;
    });

File: contrail-core/src/main/java/org/contrail/stream/network/component/ClientComponent.java
/*
 * Copyright (C)2012 D. Plaindoux.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation; either version 2, or (at your option) any
 * later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 */

package org.contrail.stream.network.component;

import java.util.ArrayList;
import java.util.List;

import org.contrail.stream.component.ComponentNotConnectedException;
import org.contrail.stream.component.pipeline.AbstractPipelineComponent;
import org.contrail.stream.flow.DataFlow;
import org.contrail.stream.network.component.flow.client.ClientDownStreamDataFlow;
import org.contrail.stream.network.component.flow.client.ClientUpStreamDataFlow;
import org.contrail.stream.network.packet.Packet;

/**
 * <code>ClientComponent</code>
 * 
 * @author Didier Plaindoux
 * @version 1.0
 */
public class ClientComponent extends AbstractPipelineComponent<Packet, Packet, Packet, Packet> {

	private final String endPoint;
	private final DataFlow<Packet> upStreamDataFlow;
	private final DataFlow<Packet> downStreamDataFlow;
	private final List<String> destinations;
	private final ArrayList<String> sources;

	public ClientComponent(String endPoint) {
		this.endPoint = endPoint;

		this.upStreamDataFlow = new ClientUpStreamDataFlow(this);
		this.downStreamDataFlow = new ClientDownStreamDataFlow(this);

		this.destinations = new ArrayList<String>();
		this.sources = new ArrayList<String>();
	}

	public String getEndPoint() {
		return endPoint;
	}

	public synchronized void addSourceId(String sourceId) {
		if (!this.acceptSourceId(sourceId)) {
			this.sources.add(sourceId);
		}
	}

	public synchronized boolean acceptSourceId(String identifier) {
		return this.sources.contains(identifier);
	}

	public synchronized void addDestinationId(String identifier) {
		if (!this.acceptDestinationId(identifier)) {
			this.destinations.add(identifier);
		}
	}

	public boolean acceptDestinationId(String identifier) {
		return this.destinations.contains(identifier);
	}

	@Override
	public DataFlow<Packet> getUpStreamDataFlow() throws ComponentNotConnectedException {
		return this.upStreamDataFlow;
	}

	@Override
	public DataFlow<Packet> getDownStreamDataFlow() throws ComponentNotConnectedException {
		return this.downStreamDataFlow;
	}
}
